File: crates/viben-core/src/cli/output.rs
//! CLI output utilities

use serde::Serialize;

/// Standard success response for JSON output
#[derive(Serialize)]
pub struct SuccessResponse<T: Serialize> {
    pub success: bool,
    pub data: T,
}

impl<T: Serialize> SuccessResponse<T> {
    pub fn new(data: T) -> Self {
        Self {
            success: true,
            data,
        }
    }
}

/// Standard error response for JSON output
#[derive(Serialize)]
pub struct ErrorResponse {
    pub success: bool,
    pub error: ErrorDetail,
}

#[derive(Serialize)]
pub struct ErrorDetail {
    pub message: String,
    pub code: Option<String>,
}

impl ErrorResponse {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            success: false,
            error: ErrorDetail {
                message: message.into(),
                code: None,
            },
        }
    }

    pub fn with_code(message: impl Into<String>, code: impl Into<String>) -> Self {
        Self {
            success: false,
            error: ErrorDetail {
                message: message.into(),
                code: Some(code.into()),
            },
        }
    }
}

/// Print JSON output to stdout
pub fn print_json<T: Serialize>(value: &T) {
    if let Ok(json) = serde_json::to_string_pretty(value) {
        println!("{}", json);
    }
}

/// Print a success message
pub fn print_success(message: &str) {
    println!("{}", message);
}

/// Print an error message
pub fn print_error(message: &str) {
    eprintln!("Error: {}", message);
}

/// Print a warning message
pub fn print_warning(message: &str) {
    eprintln!("Warning: {}", message);
}

/// Print a table header
pub fn print_table_header(columns: &[(&str, usize)]) {
    let header: Vec<String> = columns
        .iter()
        .map(|(name, width)| format!("{:width$}", name, width = width))
        .collect();
    println!("{}", header.join("  "));
    let separator: Vec<String> = columns
        .iter()
        .map(|(_, width)| "-".repeat(*width))
        .collect();
    println!("{}", separator.join("  "));
}

/// Print a table row
pub fn print_table_row(values: &[(&str, usize)]) {
    let row: Vec<String> = values
        .iter()
        .map(|(value, width)| {
            if value.len() > *width {
                format!("{}...", &value[..(width - 3)])
            } else {
                format!("{:width$}", value, width = width)
            }
        })
        .collect();
    println!("{}", row.join("  "));
}

/// Print a simple table with auto-calculated column widths
pub fn print_simple_table(headers: &[&str], rows: &[Vec<String>]) {
    // Calculate column widths
    let mut widths: Vec<usize> = headers.iter().map(|h| h.len()).collect();
    for row in rows {
        for (i, cell) in row.iter().enumerate() {
            if i < widths.len() {
                widths[i] = widths[i].max(cell.len());
            }
        }
    }

    // Print header
    let header_line: Vec<String> = headers
        .iter()
        .enumerate()
        .map(|(i, h)| format!("{:width$}", h, width = widths[i]))
        .collect();
    println!("{}", header_line.join("  "));

    // Print separator
    let separator: Vec<String> = widths.iter().map(|w| "-".repeat(*w)).collect();
    println!("{}", separator.join("  "));

    // Print rows
    for row in rows {
        let row_line: Vec<String> = row
            .iter()
            .enumerate()
            .map(|(i, cell)| {
                let width = widths.get(i).copied().unwrap_or(cell.len());
                format!("{:width$}", cell, width = width)
            })
            .collect();
        println!("{}", row_line.join("  "));
    }
}

/// Trait for types that can be displayed in a table
pub trait TableDisplay {
    /// Get the headers for the table
    fn headers() -> Vec<&'static str>;
    /// Get the row values for this item
    fn row(&self) -> Vec<String>;
}

File: packages/core/src/cli/lib/output.ts
/**
 * CLI output utilities
 */
import chalk from "chalk";
import type { CliResponse, OutputContext } from "../types";

/**
 * Create a success response
 */
export function successResponse<T>(data: T): CliResponse<T> {
  return {
    success: true,
    data,
  };
}

/**
 * Create an error response
 */
export function errorResponse(code: string, message: string): CliResponse<never> {
  return {
    success: false,
    error: {
      code,
      message,
    },
  };
}

/**
 * Output a response in the appropriate format
 * @param ctx - Output context
 * @param response - The response to output
 * @param humanRender - Optional function to render human-readable output
 */
export function output<T>(
  ctx: OutputContext,
  response: CliResponse<T>,
  humanRender?: () => void
): void {
  if (ctx.json) {
    console.log(JSON.stringify(response, null, 2));
    return;
  }

  if (ctx.quiet && response.success) {
    return;
  }

  if (humanRender) {
    humanRender();
  } else if (response.success && response.data !== undefined) {
    // Default human render - just stringify the data
    if (typeof response.data === "string") {
      console.log(response.data);
    } else {
      console.log(JSON.stringify(response.data, null, 2));
    }
  }
}

/**
 * Output an error message
 */
export function outputError(ctx: OutputContext, code: string, message: string): void {
  if (ctx.json) {
    console.log(JSON.stringify(errorResponse(code, message), null, 2));
    return;
  }

  console.error(chalk.red(`Error: ${message}`));
  if (ctx.verbose) {
    console.error(chalk.gray(`Code: ${code}`));
  }
}

/**
 * Output a success message
 */
export function outputSuccess(ctx: OutputContext, message: string): void {
  if (ctx.json) {
    console.log(JSON.stringify(successResponse({ message }), null, 2));
    return;
  }

  if (!ctx.quiet) {
    console.log(chalk.green(message));
  }
}

/**
 * Output a warning message
 */
export function outputWarning(ctx: OutputContext, message: string): void {
  if (ctx.json) {
    return; // Warnings are not included in JSON output
  }

  if (!ctx.quiet) {
    console.warn(chalk.yellow(`Warning: ${message}`));
  }
}

/**
 * Output an info message
 */
export function outputInfo(ctx: OutputContext, message: string): void {
  if (ctx.json) {
    return; // Info messages are not included in JSON output
  }

  if (!ctx.quiet && ctx.verbose) {
    console.log(chalk.blue(message));
  }
}

/**
 * Simple table output
 */
export function outputTable(
  ctx: OutputContext,
  headers: string[],
  rows: (string | number | boolean | undefined)[][]
): void {
  if (ctx.json) {
    return; // Tables are handled by the main output function
  }

  if (ctx.quiet) {
    return;
  }

  if (rows.length === 0) {
    console.log(chalk.gray("No data"));
    return;
  }

  // Calculate column widths
  const widths = headers.map((h, i) => {
    const dataWidth = Math.max(...rows.map((r) => String(r[i] ?? "").length));
    return Math.max(h.length, dataWidth);
  });

  // Print header
  const headerLine = headers
    .map((h, i) => h.padEnd(widths[i]))
    .join("  ");
  console.log(chalk.bold(headerLine));

  // Print separator
  const separator = widths.map((w) => "─".repeat(w)).join("──");
  console.log(chalk.gray(separator));

  // Print rows
  for (const row of rows) {
    const line = row
      .map((cell, i) => String(cell ?? "").padEnd(widths[i]))
      .join("  ");
    console.log(line);
  }
}

/**
 * Output a list with bullets
 */
export function outputList(
  ctx: OutputContext,
  items: string[],
  bullet = "•"
): void {
  if (ctx.json) {
    return;
  }

  if (ctx.quiet) {
    return;
  }

  for (const item of items) {
    console.log(`  ${bullet} ${item}`);
  }
}

/**
 * Output key-value pairs
 */
export function outputKeyValue(
  ctx: OutputContext,
  pairs: Record<string, string | number | boolean | undefined>
): void {
  if (ctx.json) {
    return;
  }

  if (ctx.quiet) {
    return;
  }

  const maxKeyLength = Math.max(...Object.keys(pairs).map((k) => k.length));

  for (const [key, value] of Object.entries(pairs)) {
    const paddedKey = key.padEnd(maxKeyLength);
    console.log(`${chalk.gray(paddedKey)}  ${value ?? chalk.gray("(not set)")}`);
  }
}
